diff --git a/.vscode/c_cpp_properties.json b/.vscode/c_cpp_properties.json
new file mode 100644
index 0000000..4039bef
--- /dev/null
+++ b/.vscode/c_cpp_properties.json
@@ -0,0 +1,16 @@
+{
+    "configurations": [
+        {
+            "name": "Linux",
+            "includePath": [
+                "${workspaceFolder}/**"
+            ],
+            "defines": [],
+            "compilerPath": "/usr/bin/gcc",
+            "cStandard": "c17",
+            "cppStandard": "gnu++17",
+            "intelliSenseMode": "linux-gcc-x64"
+        }
+    ],
+    "version": 4
+}
\ No newline at end of file
diff --git a/Makefile b/Makefile
index cac799d..65a91c4 100644
--- a/Makefile
+++ b/Makefile
@@ -154,7 +154,9 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
-
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_find\
 
 ifeq ($(LAB),syscall)
 UPROGS += \
diff --git a/kernel/main.c b/kernel/main.c
index 8a3dc2e..2d6ae1d 100644
--- a/kernel/main.c
+++ b/kernel/main.c
@@ -9,6 +9,7 @@ volatile static int started = 0;
 // start() jumps here in supervisor mode on all CPUs.
 void main() {
   if (cpuid() == 0) {
+    printf("[210810401] enter main, init kernel\n");
     // consoleinit();
     // printfinit();
     printf("\n");
diff --git a/kernel/proc.c b/kernel/proc.c
index 1607145..1079977 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -183,6 +183,7 @@ uchar initcode[] = {0x17, 0x05, 0x00, 0x00, 0x13, 0x05, 0x45, 0x02, 0x97, 0x05,
 
 // Set up first user process.
 void userinit(void) {
+  printf("[210810401] enter userinit\n");
   struct proc *p;
 
   p = allocproc();
@@ -197,6 +198,7 @@ void userinit(void) {
   p->trapframe->epc = 0;      // user program counter
   p->trapframe->sp = PGSIZE;  // user stack pointer
 
+  printf("[210810401] copy initcode to first user process\n");
   safestrcpy(p->name, "initcode", sizeof(p->name));
   p->cwd = namei("/");
 
diff --git a/kernel/start.c b/kernel/start.c
index f704fee..c3f062f 100644
--- a/kernel/start.c
+++ b/kernel/start.c
@@ -52,6 +52,7 @@ void start() {
     // init uart and printf
     consoleinit();
     printfinit();
+    printf("[210810401] in start, init driver, interrupts and change mode\n");
   }
 
   // switch to supervisor mode and jump to main().
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..c6af854
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,68 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "kernel/fs.h"
+#include "user.h"
+
+void find(char *path, char *fn) {
+  // 存储临时路径的数组buf和指针p
+  char buf[512], *p;
+  // 文件描述符（File Descriptor），它用于打开文件或目录并执行与该文件描述符相关的操作
+  int fd;
+  // 目录项（Directory Entry），用于存储目录中的每个文件或子目录的信息
+  struct dirent de;
+  // 文件或目录的元数据信息
+  struct stat st;
+
+  if ((fd = open(path, 0)) < 0) {
+    fprintf(2, "ls: cannot open %s\n", path);
+    return;
+  }
+
+  if (fstat(fd, &st) < 0) {
+    fprintf(2, "ls: cannot stat %s\n", path);
+    close(fd);
+    return;
+  }
+
+  // buf存储临时路径
+  strcpy(buf, path);
+  p = buf + strlen(buf);
+  *p++ = '/';
+  // 循环读取每个目录项
+  while (read(fd, &de, sizeof(de)) == sizeof(de)) {
+    // 将目录项的名称补充到路径上
+    memmove(p, de.name, DIRSIZ);
+    p[DIRSIZ] = 0;
+    if (stat(buf, &st) < 0) {
+      printf("ls: cannot stat %s\n", buf);
+      continue;
+    }
+    // 如果目录项是文件，则直接比较文件名是否相等；如果是路径，则递归调用find函数
+    switch (st.type) {
+      case T_FILE:
+        if (strcmp(fn, de.name) == 0) {
+          printf("%s\n", buf);
+        }
+        break;
+      case T_DIR:
+        if (strcmp(de.name, ".") == 0 || strcmp(de.name, "..") == 0) {
+          continue;
+        } else {
+          find(buf, fn);
+        }
+    }
+  }
+  close(fd);
+}
+
+int main(int argc, char *argv[]) {
+  if (argc != 3) {
+    printf("find needs two argument!\n");
+    exit(-1);
+  }
+  char *path = argv[1];
+  // 文件名file name
+  char *fn = argv[2];
+  find(path, fn);
+  exit(0);
+}
\ No newline at end of file
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..441b24b
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,37 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int main(int argc, char *argv[]) {
+  int pipe_to_child[2];
+  int pipe_to_father[2];
+  pipe(pipe_to_child);         // 父进程传给子进程
+  pipe(pipe_to_father);        // 子进程传给父进程
+  if (fork() == 0) {           // 如果是子进程
+    close(pipe_to_child[1]);   // 关闭写端
+    close(pipe_to_father[0]);  // 关闭读端
+    char buffer[4];
+    if (read(pipe_to_child[0], buffer, 4) > 0) {
+      printf("%d: received ping\n", getpid());
+      close(pipe_to_child[0]);
+    }
+    if (write(pipe_to_father[1], "pong", 4) != 4) {
+      printf("write error!\n");
+      exit(-1);
+    }
+    close(pipe_to_father[1]);  // 关闭写端
+  } else {                     // 如果是父进程
+    close(pipe_to_child[0]);   // 关闭读端
+    close(pipe_to_father[1]);  // 关闭写端
+    if (write(pipe_to_child[1], "ping", 4) != 4) {
+      printf("write error!\n");
+      exit(-1);
+    }
+    close(pipe_to_child[1]);  // 关闭写端
+    char buffer[4];
+    if (read(pipe_to_father[0], buffer, 4) > 0) {
+      printf("%d: received pong\n", getpid());
+      close(pipe_to_father[0]);
+    }
+  }
+  exit(0);
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..25593b4
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,13 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int main(int argc, char *argv[]) {
+  if (argc != 2) {
+    printf("Sleep needs one argument!\n");
+    exit(-1);
+  }
+  int ticks = atoi(argv[1]);
+  sleep(ticks);
+  printf("(nothing happens for a little while)\n");
+  exit(0);
+}
\ No newline at end of file
